### PR TITLE
Al/BE-130. Fix controller tests.

### PR DESCRIPTION
**Pull request by lut1k**
Fixed unit-tests and mock file in the pallet controller.
Added comments to the code in controller pallet.
Added PoolUserData to genesis block (ALICE - DOT - collateral: true).

### CODE REVIEW

File: pallets/controller/src/mock.rs [hunk 1]
```diff
@@ -177,234 +177,106 @@ impl Default for ExtBuilder {
...
+	pub fn pool_mock(mut self, pool_id: CurrencyId) -> Self {
 		self.pools.push((
-			CurrencyId::ETH,
+			pool_id,
 			Pool {
 				current_interest_rate: Rate::from_inner(0),
 				total_borrowed: Balance::zero(),
-				borrow_index: Rate::saturating_from_rational(1, 1),
-				current_exchange_rate: Rate::saturating_from_rational(8, 10),
+				borrow_index: Rate::saturating_from_rational(2, 1),
```
Comment (Apollo50) [inline pallets/controller/src/mock.rs:238]: Why did you changed borrow_index from (1, 1) to (2, 1) ? If it's initial params borrow rate must be equal 1,1

File: pallets/controller/src/tests.rs [hunk 2]
```diff
@@ -69,163 +101,207 @@ fn accrue_interest_should_not_work() {
...
+		// Overflow in calculation: kink * multiplier_per_block = 1.01 * max_value()
+		assert_noop!(
+			Controller::calculate_borrow_interest_rate(CurrencyId::KSM, 1, 200, 8),
+			Error::<Runtime>::NumOverflow
+		);
+	});
+}
+
+#[test]
+fn calculate_borrow_interest_rate_fails_if_overflow_add_baser_rate_per_block() {
```
Comment (Apollo50) [inline pallets/controller/src/tests.rs:245]: "_add_baser_rate_"  *base

File: pallets/controller/src/tests.rs [hunk 3]
```diff
@@ -69,163 +101,207 @@ fn accrue_interest_should_not_work() {
...
+				dollars(106_000_u128),
+				dollars(54_000_u128),
+				dollars(33_000_u128)
+			),
+			Ok(Rate::from_inner(3_826_771_653))
 		);
 
-		// utilization rate larger or equal than kink
+		// Utilization rate larger than kink:
+		// utilization_rate = 90000 / (18 - 8 + 90) = 0.9
```
Comment (Apollo50) [inline pallets/controller/src/tests.rs:215]: better 90 instead 90000

**Review by Apollo50 — DISMISSED** [review]

File: pallets/controller/src/mock.rs (hunk 1 above)
Comment (lut1k) [reply pallets/controller/src/mock.rs:238]: I changed this setting to test math in functions.

File: pallets/controller/src/tests.rs (hunk 3 above)
Comment (lut1k) [reply pallets/controller/src/tests.rs:215]: Fixed.

File: pallets/controller/src/tests.rs (hunk 2 above)
Comment (lut1k) [reply pallets/controller/src/tests.rs:245]: Fixed

File: pallets/controller/src/lib.rs [hunk 4]
```diff
@@ -484,25 +521,23 @@ impl<T: Trait> Module<T> {
 				.checked_mul(&oracle_price)
 				.ok_or(Error::<T>::NumOverflow)?;
 
-			if <LiquidityPools<T>>::check_user_available_collateral(&account, underlying_asset) {
+			if (<LiquidityPools<T>>::check_user_available_collateral(&account, underlying_asset) && borrow_amount > 0)
```
Comment (Pavel-Blaize) [inline pallets/controller/src/lib.rs:524]: Here can be missed the case, when the user redeems actual collateral while the borrow is active.
Could you please fix it to the next way:
```
if ( check_user_available_collateral() ) {
    sum_collateral += tokens_to_denom * m_token_balance
}
sum_borrow_plus_effects += oracle_price * borrow_balance
```
And you can remove redundant "borrow_amount > 0" and "redeem_amount > 0"

Comment (lut1k) [reply pallets/controller/src/lib.rs:524]: If we do this, then we will ban making a burning operation after the deposit operation.

Comment (Pavel-Blaize) [reply pallets/controller/src/lib.rs:524]: Also, to keep it correct add ensure that, borrow_amount and redeem_amount can be greater than 0 simultaneously.

Comment (Pavel-Blaize) [reply pallets/controller/src/lib.rs:524]: Also, in order to solve the issue with incorrect behaviour for the non-collateral redeement, redeem_allowed() should call hypotetical...() only for collaterals

Comment (lut1k) [reply pallets/controller/src/lib.rs]: Ok, fixed. Check please.

File: pallets/liquidity-pools/src/lib.rs [hunk 5]
```diff
@@ -133,6 +133,16 @@ impl<T: Trait> Module<T> {
 		});
 		Ok(())
 	}
+
+	pub fn enable_as_collateral_internal(who: &T::AccountId, pool_id: CurrencyId) -> DispatchResult {
+		PoolUserDates::<T>::mutate(who, pool_id, |p| p.collateral = true);
+		Ok(())
+	}
+
+	pub fn disable_collateral_internal(who: &T::AccountId, pool_id: CurrencyId) -> DispatchResult {
```
Comment (Pavel-Blaize) [inline pallets/liquidity-pools/src/lib.rs:142]: The user cannot just turn off the collateral.
User deposits DOTs (collateral)
User borrows some cash
User turns off the collateral for DOTs
Free redeement + free money from borrow (which is not allowed).

Please add the hypotetical...() check like the user wants to redeem the whole amount of the tokens un-marked from being collateral.

File: pallets/liquidity-pools/src/tests.rs [hunk 6]
```diff
@@ -83,3 +83,23 @@ fn update_state_on_repay_should_work() {
...
+	ExtBuilder::default().build().execute_with(|| {
+		// Alice enable as collateral DOT pool
+		assert_ok!(LiquidityPools::enable_as_collateral_internal(&ALICE, CurrencyId::DOT));
+
+		assert!(<PoolUserDates<Runtime>>::get(ALICE, CurrencyId::DOT).collateral);
+	});
+}
+
+#[test]
+fn disable_collateral_internal_should_work() {
```
Comment (Pavel-Blaize) [inline pallets/liquidity-pools/src/tests.rs:98]: This should be expanded for 2 tests:
1) Simple unit test for disabling the collateral when there are no borrows
2) Integration test for the moment, when turning off is impossible due to the current borrow.

User deposits 60 DOTs (collateral)
User takes a loan for 40 DAI
Collateral disabling for DOTs is impossible
User adds 40 USDT as collateral
Collateral disabling for DOTs is possible

File: pallets/liquidity-pools/src/lib.rs (hunk 5 above)
Comment (Pavel-Blaize) [reply pallets/liquidity-pools/src/lib.rs:142]: Probably this change should be applied for the module function, not for the internal

File: pallets/minterest-protocol/src/lib.rs [hunk 7]
```diff
@@ -219,6 +227,26 @@ decl_module! {
 				Ok(())
 			})?
 		}
+
+		/// Sender allowed the assets in the pool to be used as collateral.
+		#[weight = 10_000]
+		pub fn enable_as_collateral(origin, pool_id: CurrencyId) -> DispatchResult {
+			let sender = ensure_signed(origin)?;
```
Comment (Pavel-Blaize) [inline pallets/minterest-protocol/src/lib.rs:234]: Also, probaly only the user himself can turn on/turn off the collateral flag for his funds.

File: pallets/minterest-protocol/src/tests.rs [hunk 8]
```diff
@@ -303,3 +303,41 @@ fn repay_on_behalf_should_work() {
...
+		assert_noop!(
+			MinterestProtocol::enable_as_collateral(Origin::signed(ALICE), CurrencyId::MDOT),
+			Error::<Test>::PoolNotFound
+		);
+	});
+}
+
+#[test]
+fn disable_collateral_should_work() {
+	new_test_ext().execute_with(|| {
```
Comment (Pavel-Blaize) [inline pallets/minterest-protocol/src/tests.rs:328]: See comment above for the tests

File: pallets/liquidity-pools/src/tests.rs (hunk 6 above)
Comment (Pavel-Blaize) [reply pallets/liquidity-pools/src/tests.rs:98]: These internal unit-tests can be left as they are, but protocol tests should be expanded

File: pallets/controller/src/tests.rs [hunk 9]
```diff
@@ -69,163 +101,207 @@ fn accrue_interest_should_not_work() {
...
 			assert_err!(
 				Controller::convert_from_wrapped(CurrencyId::MBTC, Balance::max_value()),
 				Error::<Runtime>::NumOverflow
 			);
 		});
 }
 
+//FIXME. Not used yet
+#[test]
+fn calculate_interest_rate_should_work() {
```
Comment (Pavel-Blaize) [inline pallets/controller/src/tests.rs:144]: Should be removed together with calculate_interest_rate() as well
We have interest)rate calculation in another function

File: pallets/controller/src/tests.rs [hunk 10]
```diff
@@ -779,11 +985,11 @@ fn set_jump_multiplier_per_block_should_work() {
 #[test]
 fn pool_not_found() {
```
Comment (Pavel-Blaize) [inline pallets/controller/src/tests.rs:986]: Can be removed - PoolNotFound is tested in pause_specific_operation_should_work()

File: pallets/liquidity-pools/src/lib.rs (hunk 5 above)
Comment (lut1k) [reply pallets/liquidity-pools/src/lib.rs:142]: Added necessary checks to the protocol pallet in external functions.

File: pallets/liquidity-pools/src/tests.rs (hunk 6 above)
Comment (lut1k) [reply pallets/liquidity-pools/src/tests.rs:98]: Added necessary checks to the protocol pallet in external functions. Fixed unit-tests.

File: pallets/controller/src/tests.rs (hunk 9 above)
Comment (lut1k) [reply pallets/controller/src/tests.rs:144]: I removed this funciton and cleaned up storage in one of the following PR.

File: pallets/controller/src/tests.rs (hunk 10 above)
Comment (lut1k) [reply pallets/controller/src/tests.rs:986]: Removed

File: pallets/minterest-protocol/src/tests.rs (hunk 8 above)
Comment (lut1k) [reply pallets/minterest-protocol/src/tests.rs:328]: Fixed

**Review by Pavel-Blaize — APPROVED** [review]

**Pull request merged**